R/table.R

```diff
@@ -0,0 +1,105 @@
+#' Generics for tabular vectors
+#'
+#' @description
+#'
+#' \Sexpr[results=rd, stage=render]{vctrs:::lifecycle("experimental")}
+#'
+#' These generics make it possible to work with tables, i.e. vectors
+#' of dimension 2 like data frames and matrices. The main motivation
+#' of this API is to support colwise operations such as slicing
+#' columns. Operating across columns requires slightly different
+#' semantics and concepts than rowwise operations:
+#'
+#' - A tabular prototype is a zero-rows, zero-cols table. In order for
+#'   your tabular class to be compatible with vctrs, it needs to
+#'   support such empty instances.
+#'
+#' - The vector type of a data frame includes its columns (their names
+#'   and their types). The tabular type of a data frame doesn't. Hence
+#'   `tbl_ptype2()` and `tbl_cast()` accept inputs of any shape,
+#'   whereas `vec_ptype2()` and `vec_cast()` require congruent shapes
+#'   and names.
+#'
+#' The details and theory of tabular operations are in development and
+#' likely to change in the future.
+#'
+#' @param x A tabular vector.
+#'
+#' @export
+tbl_is <- function(x) {
+  .Call(vctrs_tbl_is, x)
+}
+#' @rdname tbl_is
+#' @inheritParams vec_assert
+#' @export
+tbl_assert <- function(x, arg = as_label(substitute(x))) {
+  .Call(vctrs_tbl_assert, x, arg)
+}
+
+#' @rdname tbl_is
+#' @export
+tbl_size <- function(x) {
```

Comment:
Can't we just use `vec_size()` here? Or is the size of a rectangle it's width?

Reply:
This is the width: `tbl_size(vec_cbind(x, y)) === tbl_size(x) + tbl_size(y)`.

Reply:
I think using `size` here again might be a bad idea?

Reply:
Davis suggested `rct_width()`. If we agree that keeping the API symmetric is better, this gives us the choice between `vec_size()` / `rct_size()` and `vec_height()` / `rct_width()`. In the latter case we'd have to rename `vec_size()` to `vec_height()`.